A41-xds-rbac.md

```diff
@@ -0,0 +1,199 @@
+used in non-gRPC environments as a precise service-level resource authz engine.
+However, this 1) requires another filter to perform end-user authentication (as
+TLS just provides service-level authn) and 2) does not scale as RBAC provides no
+mechanism for loading rules on-demand. Since gRPC encodes service-level
+resource identifiers in the request message, RBAC does not have access to the
+requested resource for gRPC traffic.
+
+### Related Proposals:
+
+* [A36: xDS-Enabled Servers][A36]
+* [A29: xDS-Based Security for gRPC Clients and Servers][A29]
+* [A39: xDS HTTP Filter Support][A39]
+
+[A36]: https://github.com/grpc/proposal/blob/master/A36-xds-for-servers.md
+<!-- FIXME: replace with final link once merged -->
+[A29]: https://github.com/grpc/proposal/pull/184
+[A39]: https://github.com/grpc/proposal/blob/master/A39-xds-http-filters.md
+
+## Proposal
+
+Support the [xDS RBAC HTTP filter][RBAC filter] as a registered filter type on
+xDS-enabled gRPC Servers, and support its `RBACPerRoute` configuration override.
+This does _not_ include [RBAC network filter][] support nor running the filter
+on gRPC clients. xDS v2 support is not necessary.
+
+Supporting the RBAC HTTP filter on server-side leverages [A36: xDS-Enabled
+Servers][A36] and [A39: xDS HTTP Filter Support][A39]. Server-side HTTP filters
+are less common than client-side at this point in time, so this may be the first
+usage of `RouteConfiguration` on server-side and so likely involves adding more
+complete support for A39 by creating+executing filters on server-side and
+processing `VirtualHost`s and `RoutMatch`es to determine which configuration
+should be provided to each filter. A39 should be consulted for the expected
+behavior.
+
+The core policy matching logic should be split into an "RBAC engine" to allow
+reuse with non-xDS environments. However, there is no requirement that the RBAC
+engine have a specialized API; it could simply be an interceptor.
+
+The [RBAC policy][] has many fields. All current (Envoy-implemented) fields will
+be considered in gRPC. However, some fields may have pre-determined values or
+behavior. At this time, if the `RBAC.action` is `Action.LOG` then the policy
+will be completely ignored, as if RBAC was not configurated. CEL fields are not
+supported, so if `Policy.condition` or `Policy.checked_condition` is present the
+XdsClient must NACK the configuration.
+
+The following fields of `Permission` may not be obvious how they map to gRPC:
+
+| Permission Field | gRPC Equivalent |
+| ---------------- | --------------- |
+| header           | Metadata (with caveats) |
+| url_path         | Fully-qualified RPC method name with leading slash |
+| destination_ip   | Local address for this connection |
+| destination_port | Local port for this connection |
+| metadata         | Hard-coded as empty; never matches |
+| requested_server_name | Hard-coded as empty string |
+
+Be strongly aware that Envoy Metadata has no relation to gRPC's Metadata.
```

Comment:
I suggest splitting this paragraph into two, and reordering it slightly:

"""
Because matching supports NOT, the matcher must still be processed even if a rule contains references to things that don't generally match; it is not trivial to "optimize out" the never-matching rules.

Be strongly aware that Envoy Metadata has no relation to gRPC's Metadata.  In gRPC, the term "metadata" is used to refer to request headers, whereas in Envoy it refers to internal state shared between filters.  gRPC does not currently support the latter mechanism.
"""

This way, the discussion about not optimizing out the never-matching rules (which I think is not specific to headers) comes first, and then the discussion about headers comes next.  And the two paragraphs about headers are right next to each other.

Reply:
Due to some earlier changes I had already moved the Envoy Metadata discussion down to the end. I've added a sentence giving a brief mention of what Envoy metadata is.